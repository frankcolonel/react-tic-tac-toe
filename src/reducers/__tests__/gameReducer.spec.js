import gameReducer, { initialState as state } from '../game';
import { BOARD_SIZE } from '../../reducers/gameHelpers';
import types from '../../actions/actionTypes';

describe('Game Reducer', () => {
  it('should return initial state', () => {
    expect(gameReducer()).toBe(state);
  });

  it('should choose the player', () => {
    expect(
      gameReducer(state, {
        type: types.GAME.CHOOSE_PLAYER,
        symbol: 'X',
      }),
    ).toMatchSnapshot();
  });

  it('should handle a new mark', () => {
    expect(
      gameReducer(state, {
        type: types.NEW_MARK,
        column: 0,
        row: 1,
      }),
    ).toMatchSnapshot();
  });

  it('should define player 1 as winner on vertical board spaces', () => {
    expect(
      gameReducer({
        ...state,
        boardGrid: [
          [
            1,
            null,
            null,
          ],
          [
            1,
            null,
            null,
          ],
          [
            1,
            null,
            null,
          ],
        ],
      }, {
        type: types.GAME.VERIFY,
      }),
    ).toMatchSnapshot();
  });

  it('should define player 1 as winner on diagonal board spaces', () => {
    expect(
      gameReducer({
        ...state,
        boardGrid: [
          [
            1,
            null,
            null,
          ],
          [
            null,
            1,
            null,
          ],
          [
            null,
            null,
            1,
          ],
        ],
      }, {
        type: types.GAME.VERIFY,
      }),
    ).toMatchSnapshot();
  });

  it('should define game as tie', () => {
    expect(
      gameReducer({
        ...state,
        marksCount: BOARD_SIZE ** 2,
      }, {
        type: types.GAME.VERIFY,
      }),
    ).toMatchSnapshot();
  });

  it('should change to the next player', () => {
    expect(
      gameReducer(state, {
        type: types.GAME.NEXT_PLAYER,
      }),
    ).toMatchSnapshot();
  });

  it('should restart to the initial state but keep the chosen player symbol', () => {
    expect(
      gameReducer({
        ...state,
        playersSymbols: {
          1: 'X',
          2: 'O',
        },
      }, {
        type: types.GAME.RESTART,
      }),
    ).toMatchSnapshot();
  });
});
