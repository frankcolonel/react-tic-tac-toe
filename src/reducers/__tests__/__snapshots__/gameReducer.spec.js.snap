exports[`Game Reducer should change to the next player 1`] = `
Object {
  "boardGrid": Array [
    Array [
      null,
      null,
      null,
    ],
    Array [
      null,
      null,
      null,
    ],
    Array [
      null,
      null,
      null,
    ],
  ],
  "currentPlayer": 2,
  "isTie": false,
  "marksCount": 0,
  "playersSymbols": Object {},
  "winnerMarks": Object {},
  "winnerPlayer": null,
}
`;

exports[`Game Reducer should choose the player 1`] = `
Object {
  "boardGrid": Array [
    Array [
      null,
      null,
      null,
    ],
    Array [
      null,
      null,
      null,
    ],
    Array [
      null,
      null,
      null,
    ],
  ],
  "currentPlayer": 1,
  "isTie": false,
  "marksCount": 0,
  "playersSymbols": Object {
    "1": "X",
    "2": "O",
  },
  "winnerMarks": Object {},
  "winnerPlayer": null,
}
`;

exports[`Game Reducer should define game as tie 1`] = `
Object {
  "boardGrid": Array [
    Array [
      null,
      null,
      null,
    ],
    Array [
      null,
      null,
      null,
    ],
    Array [
      null,
      null,
      null,
    ],
  ],
  "currentPlayer": 1,
  "isTie": true,
  "marksCount": 9,
  "playersSymbols": Object {},
  "winnerMarks": Object {},
  "winnerPlayer": null,
}
`;

exports[`Game Reducer should define player 1 as winner on diagonal board spaces 1`] = `
Object {
  "boardGrid": Array [
    Array [
      1,
      null,
      null,
    ],
    Array [
      null,
      1,
      null,
    ],
    Array [
      null,
      null,
      1,
    ],
  ],
  "currentPlayer": 1,
  "isTie": false,
  "marksCount": 0,
  "playersSymbols": Object {},
  "winnerMarks": Object {
    "0": Array [
      0,
    ],
    "1": Array [
      1,
    ],
    "2": Array [
      2,
    ],
  },
  "winnerPlayer": 1,
}
`;

exports[`Game Reducer should define player 1 as winner on vertical board spaces 1`] = `
Object {
  "boardGrid": Array [
    Array [
      1,
      null,
      null,
    ],
    Array [
      1,
      null,
      null,
    ],
    Array [
      1,
      null,
      null,
    ],
  ],
  "currentPlayer": 1,
  "isTie": false,
  "marksCount": 0,
  "playersSymbols": Object {},
  "winnerMarks": Object {
    "0": Array [
      0,
    ],
    "1": Array [
      0,
    ],
    "2": Array [
      0,
    ],
  },
  "winnerPlayer": 1,
}
`;

exports[`Game Reducer should handle a new mark 1`] = `
Object {
  "boardGrid": Array [
    Array [
      null,
      1,
      null,
    ],
    Array [
      null,
      null,
      null,
    ],
    Array [
      null,
      null,
      null,
    ],
  ],
  "currentPlayer": 1,
  "isTie": false,
  "marksCount": 1,
  "playersSymbols": Object {},
  "winnerMarks": Object {},
  "winnerPlayer": null,
}
`;

exports[`Game Reducer should restart to the initial state but keep the chosen player symbol 1`] = `
Object {
  "boardGrid": Array [
    Array [
      null,
      null,
      null,
    ],
    Array [
      null,
      null,
      null,
    ],
    Array [
      null,
      null,
      null,
    ],
  ],
  "currentPlayer": 1,
  "isTie": false,
  "marksCount": 0,
  "playersSymbols": Object {
    "1": "X",
    "2": "O",
  },
  "winnerMarks": Object {},
  "winnerPlayer": null,
}
`;
